### PR TITLE
Hint at the list of conversions in FFI exercise

### PR DESCRIPTION
**Pull request by mgeisler**
I see people struggle a lot with guessing why they need to convert between all these types. The explanations here should help with that.

### CODE REVIEW

File: src/exercises/day-3/safe-ffi-wrapper.md [hunk 1]
```diff
@@ -10,13 +10,36 @@ You will want to consult the manual pages:
...
+You will convert between all these types:
+
+- `&str` to `CString`: you need to allocate space for a trailing `\0` character,
+- `CString` to `*const i8`: you need a pointer to call C functions,
+- `*const i8` to `&CStr`: you need something which can find the trailing `\0` character,
+- `&CStr` to `&[u8]`: you will ultimately return a slice of bytes,
+- `&[u8]` to `&OsStr`: you can create an `OsString` from `&[u8]`. Use
+  [`OsStrExt`](https://doc.rust-lang.org/std/os/unix/ffi/trait.OsStrExt.html)
+  for this.
+- `&OsStr` to `OsString`: you need to take ownership of the data before you call
```
Comment (sakex) [inline src/exercises/day-3/safe-ffi-wrapper.md:31]: Usually, it's a copy

File: src/exercises/day-3/safe-ffi-wrapper.md [hunk 2]
```diff
@@ -10,13 +10,36 @@ You will want to consult the manual pages:
...
+|----------------------------|----------------|--------------------------------|
+| [`str`] and [`String`]     | UTF-8          | Text processing in Rust        |
+| [`CStr`] and [`CString`]   | NUL-terminated | Communicating with C functions |
+| [`OsStr`] and [`OsString`] | OS-specific    | Communicating with the OS      |
+
+You will convert between all these types:
+
+- `&str` to `CString`: you need to allocate space for a trailing `\0` character,
+- `CString` to `*const i8`: you need a pointer to call C functions,
+- `*const i8` to `&CStr`: you need something which can find the trailing `\0` character,
```
Comment (sakex) [inline src/exercises/day-3/safe-ffi-wrapper.md:26]: `*const i8` to `&[u8]` might be useful for other encodings/to write your own abstractions

**Review by sakex — APPROVED** [review]

File: src/exercises/day-3/safe-ffi-wrapper.md (hunk 1 above)
Comment (mgeisler) [reply src/exercises/day-3/safe-ffi-wrapper.md:31]: What is a copy? In Rust terms, think it's more precise to say that you get an `OsString` from a `&OsStr` by doing a heap allocation (ownership!) and then copying the data into the new place on the heap.

File: src/exercises/day-3/safe-ffi-wrapper.md (hunk 2 above)
Comment (mgeisler) [reply src/exercises/day-3/safe-ffi-wrapper.md:26]: Right, definitely and that's what you get if you go `*const i8` to `&CStr` to `&[u8]` which is the next line. I think we agree :smile:

**Comment by sakex** [comment]
I think it's wrong to say "you need to take ownership of the data before you call". You don't take ownership of the data. You make an owned copy.On Apr 27, 2023 23:43, Martin Geisler ***@***.***> wrote:
@mgeisler commented on this pull request.



In src/exercises/day-3/safe-ffi-wrapper.md:
> +| Types                      | Encoding       | Use                            |
+|----------------------------|----------------|--------------------------------|
+| [`str`] and [`String`]     | UTF-8          | Text processing in Rust        |
+| [`CStr`] and [`CString`]   | NUL-terminated | Communicating with C functions |
+| [`OsStr`] and [`OsString`] | OS-specific    | Communicating with the OS      |
+
+You will convert between all these types:
+
+- `&str` to `CString`: you need to allocate space for a trailing `\0` character,
+- `CString` to `*const i8`: you need a pointer to call C functions,
+- `*const i8` to `&CStr`: you need something which can find the trailing `\0` character,
+- `&CStr` to `&[u8]`: you will ultimately return a slice of bytes,
+- `&[u8]` to `&OsStr`: you can create an `OsString` from `&[u8]`. Use
+  [`OsStrExt`](https://doc.rust-lang.org/std/os/unix/ffi/trait.OsStrExt.html)
+  for this.
+- `&OsStr` to `OsString`: you need to take ownership of the data before you call

What is a copy? In Rust terms, think it's more precise to say that you get an OsString from a &OsStr by doing a heap allocation (ownership!) and then copying the data into the new place on the heap.

**Comment by sakex** [comment]
Or "clone" might sound better than copy .On Apr 28, 2023 00:27, ***@***.*** wrote:I think it's wrong to say "you need to take ownership of the data before you call". You don't take ownership of the data. You make an owned copy.On Apr 27, 2023 23:43, Martin Geisler ***@***.***> wrote:
@mgeisler commented on this pull request.



In src/exercises/day-3/safe-ffi-wrapper.md:
> +| Types                      | Encoding       | Use                            |
+|----------------------------|----------------|--------------------------------|
+| [`str`] and [`String`]     | UTF-8          | Text processing in Rust        |
+| [`CStr`] and [`CString`]   | NUL-terminated | Communicating with C functions |
+| [`OsStr`] and [`OsString`] | OS-specific    | Communicating with the OS      |
+
+You will convert between all these types:
+
+- `&str` to `CString`: you need to allocate space for a trailing `\0` character,
+- `CString` to `*const i8`: you need a pointer to call C functions,
+- `*const i8` to `&CStr`: you need something which can find the trailing `\0` character,
+- `&CStr` to `&[u8]`: you will ultimately return a slice of bytes,
+- `&[u8]` to `&OsStr`: you can create an `OsString` from `&[u8]`. Use
+  [`OsStrExt`](https://doc.rust-lang.org/std/os/unix/ffi/trait.OsStrExt.html)
+  for this.
+- `&OsStr` to `OsString`: you need to take ownership of the data before you call

What is a copy? In Rust terms, think it's more precise to say that you get an OsString from a &OsStr by doing a heap allocation (ownership!) and then copying the data into the new place on the heap.

File: src/exercises/day-3/safe-ffi-wrapper.md (hunk 1 above)
Comment (mgeisler) [reply src/exercises/day-3/safe-ffi-wrapper.md:31]: > You don't take ownership of the data. You make an owned copy

Ah... now I get what you mean! You're right that "the data" is a copy after you construct the `OsString`. Let me make a follwup PR.

Comment (mgeisler) [reply src/exercises/day-3/safe-ffi-wrapper.md:31]: #602 has more changes to this. Please give is a look :smile:

**Pull request merged**